File: ruby/benchmark.rb
require "benchmark"

if(ARGV.size != 1)
  puts "Usage: ruby benchmark.rb <filename>"
  exit 1
end

def measure(data, pattern)
  count = 0
  elapsed = Benchmark.measure {
    count = data.scan(pattern).size
  }

  puts "#{elapsed.real * 1e3} - #{count}"
end

data = File.read(ARGV[0])

# Email
measure(data, /[\w\.+-]+@[\w\.-]+\.[\w\.-]+/)

# URI
measure(data, /[\w]+:\/\/[^\/\s?#]+[^\s?#]+(?:\?[^\s#]*)?(?:#[^\s]*)?/)

# IP
measure(data, /(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])/)

File: perl/benchmark.pl
use Time::HiRes qw(gettimeofday);

sub measure {
    my ($data, $pattern) = @_;

    my $start = Time::HiRes::gettimeofday();

    my $count = () = $data =~ /$pattern/g;

    my $elapsed = (Time::HiRes::gettimeofday() - $start) * 1e3;

    printf("%f - %d\n", $elapsed, $count);
}

if (@ARGV != 1) {
  die "Usage: ./benchmark.pl <filename>\n";
}

my ($filename) = @ARGV;

open my $fh, '<', $filename or die 'Could not open file.';
my $text;
read $fh, $data, -s $filename;
close $fh;

# Email
measure $data, '[\w\.+-]+@[\w\.-]+\.[\w\.-]+';

# URI
measure $data, '[\w]+:\/\/[^\/\s?#]+[^\s?#]+(?:\?[^\s#]*)?(?:#[^\s]*)?';

# IP
measure $data, '(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])';



